core/regexp.c

```diff
@@ -13,49 +13,65 @@ void uwsgi_opt_pcre_jit(char *opt, char *value, void *foobar) {
 #endif
 }
 
-int uwsgi_regexp_build(char *re, pcre ** pattern, pcre_extra ** pattern_extra) {
+int uwsgi_regexp_build(char *re, pcre2_code ** pattern) {
 
-	const char *errstr;
-	int erroff;
+	int errnbr;
+	long unsigned int erroff;
 
-	*pattern = pcre_compile((const char *) re, 0, &errstr, &erroff, NULL);
+	*pattern = pcre2_compile((const unsigned char *) re, PCRE2_ZERO_TERMINATED, 0, &errnbr, &erroff, NULL);
 	if (!*pattern) {
-		uwsgi_log("pcre error: %s at offset %d\n", errstr, erroff);
+		uwsgi_log("pcre error: code %d at offset %d\n", errnbr, erroff);
 		return -1;
 	}
 
-	int opt = uwsgi.pcre_jit;
-
-	*pattern_extra = (pcre_extra *) pcre_study((const pcre *) *pattern, opt, &errstr);
-	if (*pattern_extra == NULL && errstr != NULL) {
-		pcre_free(*pattern);
-		uwsgi_log("pcre (study) error: %s\n", errstr);
-		return -1;
+	if (uwsgi.pcre_jit) {
+		errnbr = pcre2_jit_compile(*pattern, PCRE2_JIT_COMPLETE);
+		if (errnbr) {
+			uwsgi_log("pcre JIT compile error code %d\n", errnbr);
+			return -1;
+		}
 	}
 
 	return 0;
 
 }
 
-int uwsgi_regexp_match(pcre * pattern, pcre_extra * pattern_extra, char *subject, int length) {
+int uwsgi_regexp_match(pcre2_code *pattern, const char *subject, int length) {
 
-	return pcre_exec((const pcre *) pattern, (const pcre_extra *) pattern_extra, subject, length, 0, 0, NULL, 0);
+	return pcre2_match(pattern, (const unsigned char *)subject, length, 0, 0, NULL, NULL);
 }
 
-int uwsgi_regexp_match_ovec(pcre * pattern, pcre_extra * pattern_extra, char *subject, int length, int *ovec, int n) {
+int uwsgi_regexp_match_ovec(pcre2_code *pattern, const char *subject, int length, int *ovec, int n) {
+
+	int rc;
+	int i;
+	pcre2_match_data *match_data;
+	size_t *pcre2_ovec;
+
+	match_data = pcre2_match_data_create_from_pattern(pattern, NULL);
+	rc = pcre2_match(pattern, (const unsigned char *)subject, length, 0, 0, match_data, NULL);
 
 	if (n > 0) {
-		return pcre_exec((const pcre *) pattern, (const pcre_extra *) pattern_extra, subject, length, 0, 0, ovec, (n + 1) * 3);
+		// copy pcre2 output vector to uwsgi output vector
+		pcre2_ovec = pcre2_get_ovector_pointer(match_data);
+		for (i=0;i<n;i++) {
+			ovec[i] = pcre2_ovec[i];
+		}
 	}
-	return pcre_exec((const pcre *) pattern, (const pcre_extra *) pattern_extra, subject, length, 0, 0, NULL, 0);
+
+	pcre2_match_data_free(match_data);
+
+	return rc;
 }
 
-int uwsgi_regexp_ovector(pcre * pattern, pcre_extra * pattern_extra) {
+int uwsgi_regexp_ovector(pcre2_code *pattern) {
 
 	int n;
+	pcre2_match_data *match_data;
 
-	if (pcre_fullinfo((const pcre *) pattern, (const pcre_extra *) pattern_extra, PCRE_INFO_CAPTURECOUNT, &n))
-		return 0;
+	match_data = pcre2_match_data_create_from_pattern(pattern, NULL);
+	n = pcre2_get_ovector_count(match_data);
+	pcre2_match_data_free(match_data);
```

Comment:
Could pcre2_pattern_info not have been used?  Which would have been a smaller change.
```C
       if (pcre2_pattern_info((const pcre2_code *) pattern, PCRE2_INFO_CAPTURECOUNT, &n))
```

Reply:
You are most probably right, will test and update accordingly.

Reply:
It crashes, this needs more time.